### PR TITLE
Encoding fixes

### PR DESCRIPTION
**Pull request by doomspork**

### CODE REVIEW

File: test/encode_test.exs [hunk 1]
```diff
@@ -276,72 +276,26 @@ defmodule AvroEx.Encode.Test do
...
-      ints =
-        items
-        |> Enum.map(fn item ->
-          {:ok, v} = @test_module.encode(int_schema, item)
-          v
-        end)
-        |> Enum.join()
-
-      assert {:ok, <<_count::8>> <> ^ints} = @test_module.encode(schema, items)
+      assert {:ok, <<6, 2, 4, 6, 0>>} = @test_module.encode(schema, [1, 2, 3])
```
Comment (doomspork) [inline test/encode_test.exs:293]: Stop using the library to test itself. This was covering up bugs.

File: lib/avro_ex/encode.ex [hunk 2]
```diff
@@ -157,12 +157,15 @@ defmodule AvroEx.Encode do
...
+          Enum.reduce(map, acc, fn {k, v}, acc ->
+            key = do_encode(%Primitive{type: :string}, context, k)
+            value = do_encode(values, context, v)
 
-          acc <> key <> value
-        end)
+            acc <> key <> value
+          end)
+
+        encoded_map <> <<0>>
```
Comment (michaelstalker) [inline lib/avro_ex/encode.ex:168]: The spec doesn't mention a 0 terminator for [maps](https://avro.apache.org/docs/current/spec.html#map_encoding) like it does with [arrays](https://avro.apache.org/docs/current/spec.html#array_encoding). Have you tested these changes to the map implementation yet?

File: lib/avro_ex/decode.ex [hunk 3]
```diff
@@ -164,37 +164,37 @@ defmodule AvroEx.Decode do
...
-      end)
-
-    {Enum.reverse(decoded_items), rest}
+    if count > 0 do
+      {decoded_items, rest} =
+        Enum.reduce(1..count, {[], buffer}, fn _, {decoded_items, buffer} ->
+          {decoded_item, buffer} = do_decode(item_schema, context, buffer)
+          {[decoded_item | decoded_items], buffer}
+        end)
+      {Enum.reverse(decoded_items), String.slice(rest, 1..-1)}
```
Comment (michaelstalker) [inline lib/avro_ex/decode.ex:176]: Why do we drop the first item of `rest`?

File: lib/avro_ex/encode.ex (hunk 2 above)
Comment (michaelstalker) [reply lib/avro_ex/encode.ex:168]: Update: Ignore the question about testing maps. You tested [here](https://github.com/rentpath/avro_ex/pull/2/files#diff-c76d23b06408448815c816c656eff2ccR281). I'm wondering why the zero is on the end, though. Do you know?

Comment (doomspork) [reply lib/avro_ex/encode.ex:168]: > Maps are encoded as a series of blocks. Each block consists of a long count value, followed by that many key/value pairs. __A block with count zero indicates the end of the map.__ Each item is encoded per the map's value schema.

File: lib/avro_ex/decode.ex (hunk 3 above)
Comment (doomspork) [reply lib/avro_ex/decode.ex:176]: That's the `<<0>>` terminal byte.

File: lib/avro_ex/encode.ex (hunk 2 above)
Comment (michaelstalker) [reply lib/avro_ex/encode.ex:168]: 💡 I hadn't understood what that part of the spec meant. I think it makes more sense now.

**Review by michaelstalker — APPROVED** [review]

**Pull request merged**